juju/juju
# 2.8 - Populate Link-Layer Device Origin from Instance Poller

## Description of change

This patch refactors the instance-poller API server logic for updating link-layer devices, so that device address origin is managed.

Observed devices and addresses have provider ID set as appropriate, observed addresses have their origin set to _provider_ and unrecognised addresses have origin set to _machine_ indicating that the provider relinquishes control of them.

The opportunity is taken here to reduce reliance on the unwieldy methods `SetParentLinkLayerDevicesBeforeTheirChildren` and `SetDevicesAddressesIdempotently`. Instead a modular `ModelOperation` approach is preferred. The hope is that these methods can be removed entirely at some point.

In the interests of keeping patch size reviewable, there are the following notable ommissions:
- We do not currently set provider subnet ID or provider network ID for addresses. It was noticed that although network ID is set in args, there is no field for it in state. This will be addressed in a follow-up patch. Omitting the subnet ID while we work to include it with the network ID is OK, since it does not break any existing functionality.
- Controller machines did not previously, and still do not handle devices correctly. This is because the machiner updates provider link-layer data via the networking common API for controllers. This results in duplicate devices on AWS, _unsupported0_ from the provider and _eth0_ from the machine, with only the provider device having a provider ID. This will be rectified in a follow-up.

## QA steps

- Bootstrap to AWS and add a machine.
- Check that the machine's main NIC (ens3 for example) gets a provider ID in the `linklayerdevices` Mongo collection.
- Check that the address for the NIC gets origin `provider` in the `ip.addresses` collection.
- Same steps as above on MAAS, but the address should also get a provider ID.

## Documentation changes

None.

## Bug reference

- Advances https://bugs.launchpad.net/bugs/1733968
- https://bugs.launchpad.net/bugs/1855263
